_posts/2021-11-23-how-to-dry-your-ref-arch.adoc

```diff
@@ -0,0 +1,594 @@
+deploy that component across your environments. We will use this inventory to identify the common configuration across
+all the deployments.
+
+Example:
+
+* `vpc`
+** `dev/us-west-2/dev/networking/vpc`
+** `stage/us-west-2/stage/networking/vpc`
+** `prod/us-west-2/prod/networking/vpc`
+* `openvpn-server`
+** `dev/us-west-2/dev/networking/openvpn-server`
+** `stage/us-west-2/stage/networking/openvpn-server`
+** `prod/us-west-2/prod/networking/openvpn-server`
+* `ecs-deploy-runner`
+** `dev/us-west-2/mgmt/ecs-deploy-runner`
+** `stage/us-west-2/mgmt/ecs-deploy-runner`
+** `prod/us-west-2/mgmt/ecs-deploy-runner`
+** `shared/us-west-2/mgmt/ecs-deploy-runner`
+** `logs/us-west-2/mgmt/ecs-deploy-runner`
+** `security/us-west-2/mgmt/ecs-deploy-runner`
+
+The changes will be isolated to Terragrunt configuration files (syntactic changes vs semantic changes). This means that
+you won't be making changes to the underlying behavior of the Terraform code or how Terraform is called. As such, there
+will be no need for rolling out the changes with `apply`, or performing any state migrations to avoid backward
+incompatibilities.
+
+Given that, the order in which the components are updated does not matter. You can update the components in whatever order
+you would like.
+
+
+[[step_2]]
+==== Step 2: Identify common configurations
+
+Once you know which component is being updated and which Terragrunt configuration files deploy that component, the next
+step is to identify the common configuration across the deployments. To do this, we will run through a diff utility to
+compare each configuration against a single reference configuration.
+
+The reason we only need to run the diff utility against a single reference point is because we are only looking for
+configurations that are common across ALL environments. Therefore, a difference in a single comparison is enough to rule
+out that configuration as a common config.
+
+Start off by choosing a reference config and generating an initial set of candidate common configurations:
+
+. Choose one of the Terragrunt configurations as your reference point. This can be arbitrary, but we recommend using `dev` as the reference point.
+. Note all the top level blocks and attributes in the reference configuration. For example, in the file https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/6ca162dd1a8d8d6b7cf05e6a22bc4ac7bf01215f/examples/for-production/infrastructure-live/dev/us-west-2/dev/networking/vpc/terragrunt.hcl[dev/us-west-2/dev/networking/vpc/terragrunt.hcl], the list would be:
+** `terraform`
+** `include`
+** `locals`
+** `inputs`
+. Also note all the top level keys in the `inputs` attribute in the reference configuration. In the same file, the list would be:
+** `vpc_name`
+** `num_nat_gateways`
+** `cidr_block`
+** `kms_key_user_iam_arns`
+** `eks_cluster_names`
+** `tag_for_use_with_eks`
+. Cross off the `include` block, as Terragrunt does not support nested include blocks.
```

Comment:
"Cross off" makes me think we should have been really clear about what list they're working against. The first time we mention a list is in the Example Inventory list with `vpc`, `openvpn-server`, and `ecs-deploy-runner`. But now I have more points to raise.

* Inventory list: currently it is a list of 3 different example inventory lists. So I feel like really it should be formatted like "Example Inventory for VPC component", "Example Inventory for OpenVPN Server component", etc. That way people are not trying to maintain a list of multiple components. You're really only working off one at a time.
* Then in this section, further up, we should reference that inventory list of the component, and say, "augment the list with the top level blocks, and the top level input keys, for each terragrunt.hcl", if that's correct. 
* Then this point about crossing off will make more sense. We're slowly removing things to check for.
* Another point of confusion for me: what we're crossing off are all the things that are _different_, so that we end up with a reference `dev` configuration that only has what's common across accounts. So the point about "as Terragrunt does not support nested include blocks" doesn't make sense to me, because I'm thinking "remove all the differences" and now I'm wondering "why does nesting matter?" Maybe we reword the reasoning to "the include block will have to be different for each environment." And for the next point "the inputs will have to be assumed different for each environment for now. We will find the common inputs later."

I hope what I'm saying makes sense.

Reply:
I reworded this section: 3a46769a05b97e52faad4bdaf8c9a7e713c3d130

Hope that makes it more clear. The key here is that you are not building up the common config just yet, only tracking what should go into the common config. So this is literally an inventory of just the blocks and attributes that need to be copied over from the reference.

Regarding `include`, I changed the wording to hopefully make it more clear - does the new wording make more sense about why we are crossing off `include` immediately?

Reply:
Thank you for making these changes! I think it's more clear now that the user is maintaining a text-file list while comparing terragrunt.hcl files. I think this works.